torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,222 @@
+    return mod
+
+def add_quant_dequant_module(module, qconfig_dict, name):
+    assert name != ''
+    qconfig = qconfig_dict[name]
+    splits = name.split('.')
+    print('splits:', splits)
+    for split in splits:
+        parent_module = module
+        module = module._modules[split]
+    assert parent_module is not module
+    parent_module._modules[splits[-1]] = QuantWrapper(module, qconfig)
+
+def add_quant_dequant(module, qconfig_dict):
+    r"""Add QuantStub and DeQuantStub module and Wrap the module in QuantWrapper
+    according to qconfig_dict
+    """
+    mod_key_set = set()
+    for name, _ in module.named_modules():
+        print('name', name)
+        dict_key = get_config_key(name, qconfig_dict)
+        if dict_key is not None:
+            mod_key_set.add(dict_key)
+
+    # We need to update the modules from inner most
+    # module to outer modules(in decreasing order of number of '.')
+    print('mod_key_set', mod_key_set)
+    mod_key_heap = []
+    for name in mod_key_set:
+        heapq.heappush(mod_key_heap, (-name.count('.'), name))
+    while mod_key_heap:
+        _, name = heapq.heappop(mod_key_heap)
+        if name == '':
+            module = QuantWrapper(module, qconfig_dict[name])
+        else:
+            add_quant_dequant_module(module, qconfig_dict, name)
+    return module
+
+def prepare(module, qconfig_dict):
+    r"""Transform a module to a quantized module according to qconfig_dict,
+    it adds observer and quant dequant stub modules to the module. and
+    changes the forward method to invoke them
+
+    Args:
+        mod: instance of the module we want to transform
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+                     configuration
+    """
+    add_observer(module, qconfig_dict)
+    return add_quant_dequant(module, qconfig_dict)
+
+class QuantStub(nn.Module):
+    r"""Quantize stub module which will be replace to actual Quantize module
+    in swap_module function
+    """
+    def __init__(self, qconfig):
+        super(QuantStub, self).__init__()
+        self.add_module('observer', qconfig.activation())
+
+    def forward(self, x):
+        self.observer(x)
+        return x
+
+class DeQuantStub(nn.Module):
+    r"""Dequantize stub module which will be replace to actual DeQuantize module
+    in swap_module function
+    """
+    def __init__(self):
+        super(DeQuantStub, self).__init__()
+
+    def forward(self, x):
+        return x
+
+def quantize(module, qconfig_dict, eval_fn, *eval_args):
+    r"""Given a float module and qconfig_dict, convert it to a quantized module
+
+    First it will prepare the module to add observer and quant
+    deqaunt stub modules, then it calls `eval_fn` which will run the calibration
+    step and observers will record the stats of the tensors, after that we will
+    call `convert` which will convert the module to a quantized module.
+    """
+    module = prepare(module, qconfig_dict)
```

Comment:
it's a bit confusing that we both return the modified module tree and mutate the original one. Not sure how to fix it consistently (we could do deepcopy, but I'm not sure it's worth it)

Reply:
the step that modifies it is `add_quant_dequant`, all other functions are modifying module inplace. I think we cannot avoid returning new module if we use `QuantWrapper` because the top level module can be replaced with `QuantWrapper(module)`, however, if we rule out this possibility, i.e. constraint user to only provide a non-terminal module as input, then we don't need to return.

Reply:
I'd add it to docstring that module would be modified in place but the new wrapper can be returned